eval.go

```diff
@@ -449,6 +456,39 @@ func (d *indexData) scoreFileMatch(fileMatch *FileMatch, doc uint32, mt matchTre
 	fileMatch.addScore("repo-rank", scoreRepoRankFactor*float64(md.Rank)/maxUInt16, opts.DebugScore)
 }
 
+// scoreFileUsingBM25 computes a score for the file match using an approximation to BM25, the most common scoring
+// algorithm for keyword search: https://en.wikipedia.org/wiki/Okapi_BM25. It implements all parts of the formula
+// except inverse document frequency (idf), since we don't have access to global term frequency statistics.
+//
+// This scoring strategy ignores all other signals including document ranks. This keeps things simple for now,
+// since BM25 is not normalized and can be tricky to combine with other scoring signals.
+func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands []*candidateMatch, opts *SearchOptions) {
```

Comment:
Why do we need `cands` here? Can we not recreate what you need from fileMatch? Having both seems redundant.

Reply:
It turned out cleaner this way, since we don't need to iterate through both line and chunk matches. Also it's a little harder to pull out the original terms from line and chunk matches, since you just have content and a list of ranges over that content.